agen.1
.TH agen 1 "October 1999" Linux "Linux Audio Manual"
.SH NAME
agen \- Software Audio Generator
.SH SYNOPSIS
.B agen [-v][-f <freq>][-l <level>][-c left|right|both][-p norm|rev]\
[-t <secs>] <dsp-dev>

.SH DESCRIPTION
\fBAgen\fP is a command-line tool for producing audio test tones.

.SH OPTIONS
.TP
.B \-v
Print a version identifier and then exit.
.TP
.B \-f <freq>
Output a tone of frequency \fI<freq>\fP Hz.  If not specified, a frequency of 
1000 Hz will be used by default.
.TP
.B \-l <level>
Output the tone at \fI<level>\fP dB below the maximum output level.  If not
specified, a level of 0 dB will be used --i.e. maximum output level.
.TP
.B \-c left|right|both
Output the tone on the corresponding channel(s).  If not specified, the tone
will be output on both channels.
.TP
.B \-p norm|rev
If given as \fBnorm\fP, the tone will be output with both channels in phase.
If \fBrev\fP, then channels will be 180 degrees out of phase.
.TP
.B \-t <secs>
Output the tone for \fI<secs>\fP seconds.  If not specified, the tone will be
output until the INT key (most often CNTL-C) is pressed.
.SH NOTES
The range of tones supported by \fBagen\fP is limited by the hardware 
capabilities of the audio \fI<device>\fP.  With modern audio hardware, 
it should be possible to output in the range 20 Hz - 20 kHz easily.
.SH TODO
\fBAgen\fP's functionality should really be merged into
.BR wavgen (1)
.
.SH AUTHOR
Fred Gleason <fredg@wava.com>
.SH SEE ALSO
.BR wavgen (1)


History
                    History file for the WAVLIB Project

-------------------------------------------------------------------------------
11/20/1999
Start of history file

Current released version is 2.2.2

Added the aproc(1) command, a command-line audio processor for wav files.
Currently, it only supports the 'normalize' function (-N), but is intended
as a catch-all for adding additional processing features as needed through
addition of additional command flags.

Added a man page for the aproc(1) command.

Fixed a bug in OpenWav() that was causing the file pointer to not be
pointing at the start of the wave data at function return with certain 
combinations of chunks. 

Fixed a typo in the ameter(1) man page.
-------------------------------------------------------------------------------

12/1/1999 -- v 2.3.1
Added a conditional definition of PI (definition is missing from glibc).
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.0

Cleaned up code in ameter(1) so that levels are now displayed in true 
logarithmic dB.

Added a '-l' option to ameter(1) to allow the user to set the meter reference
level

Added color to ameter(1), with a green and red segment.  Added a '-t' option
to allow the user to define where the border should be.

Updated the ameter(1) man page.

Fixed a bug in infowav(1) that was causing it to corrupt wav files when invoked
with the '-p' while omitting the audio-device name that customarily follows it.

Fixed a bug in infowav(1) that was causing extraneous output to be printed.

Added a '-t' option to recwav(1) to allow the user to specify a record length.

Updated the recwav(1) man page.

Added a TailTrim() API call to remove audio below a given set threshold level
from the end of an existing wav file.
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.1

Restored the RecordWavFile() API call, which somehow got dropped out of
v2.4.0.
-------------------------------------------------------------------------------

5/23/2000 -- v 2.4.2

Fixed a fencepost bug in TailTrim() that was causing false positive 
threshold detects on certain wav files.
-------------------------------------------------------------------------------

agen.c
/* agen.c
 *
 *  This plays a tone on the specified device
 *
 *   (C) Copyright 2002-2005 Fred Gleason <fredg@salemradiolabs.com>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation; either version 2 of
 *   the License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public
 *   License along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <strings.h>
#include <unistd.h>
#include <sys/soundcard.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <math.h>
#include <time.h>
#include <string.h>

#include "wavlib.h"

#define BUFFER_SIZE 16384
#define SAMPLE_RATE 44100
#define SAMPLE_SIZE 16
#define USAGE "agen [-f freq][-l level][-c left|right|both][-p now|rev][-t secs][-v] <device>\n"

int main(int argc,char *argv[])
{
  int i;
  int hAudio;
  unsigned char cBuffer[BUFFER_SIZE];
  int dRequest;
  float fFreq=1000;
  float fLevel=0;
  float fRatio;
  int dChan=0;     /* 0=Both, 1=Left, 2=Right */
  char sChan[10];
  int dPhase=0;    /* 0=Normal, 1=Reverse */
  char sPhase[10];
  int dTime=0;
  int dMatch=0;
  int dDebug=0;
  int dSampleRate;
  char sDevice[128];  /* Name of audio device */
  float fSampleRate;
  float fGain;
  float fAngle;
  unsigned long ldCount,ldTimeline=0,ldLimit=BUFFER_SIZE;
  short int dSample;
  time_t tmTimestamp,tmTime;

  if(argc<2) {
    printf(USAGE);
    exit(0);
  }

  if(argc==2) {
    if(strcmp(argv[1],"-v")==0) {
      printf("agen v");
      printf(VERNUM);
      printf("\n");
      exit(0);
    }
  }

  /* Get device name */
  if(sscanf(argv[argc-1],"%s",sDevice)!=1) {
    printf("agen: invalid audio device\n");
    exit(1);
  }

  /* Get options */
  for(i=1;i<argc-1;i++) {
    dMatch=0;
    if(strcmp(argv[i],"-v")==0) {     /* Version */
      printf("agen v");
      printf(VERNUM);
      printf("\n");
      exit(0);
    }
    if(strcmp(argv[i],"-d")==0) {     /* Debug Mode */
      dMatch=1;
      dDebug=1;
    }
    
    if(strcmp(argv[i],"-f")==0) {     /* Frequency */
      dMatch=1;
      if(sscanf(argv[++i],"%f",&fFreq)!=1) {
	printf("agen: invalid frequency\n");
	exit(1);
      }
    }
    if(strcmp(argv[i],"-l")==0) {    /* Level */
      dMatch=1;
      if(sscanf(argv[++i],"%f",&fLevel)!=1) {
	printf("agen: invalid level\n");
	exit(1);
      }
      fLevel=-fLevel;
    }
    if(strcmp(argv[i],"-t")==0) {    /* Time */
      dMatch=1;
      if(sscanf(argv[++i],"%d",&dTime)!=1) {
	printf("agen: invalid time interval\n");
	exit(1);
      }
    }
    if(strcmp(argv[i],"-c")==0) {    /* Channel */
      dMatch=1;
      if(sscanf(argv[++i],"%s",sChan)!=1) {
	printf("agen: invalid time interval\n");
	exit(1);
      }
      if(strcasecmp(sChan,"both")==0) {
	dChan=0;
      }
      else {
	if(strcasecmp(sChan,"left")==0) {
	  dChan=1;
	}
	else {
	  if(strcasecmp(sChan,"right")==0) {
	    dChan=2;
	  }
	  else {
	    printf("agen: invalid channel\n");
	    exit(1);
	  }
	}
      }  
    }
    if(strcmp(argv[i],"-p")==0) {    /* Phase */
      dMatch=1;
      if(sscanf(argv[++i],"%s",sPhase)!=1) {
	printf("agen: invalid phase setting\n");
	exit(1);
      }
      if(strcasecmp(sPhase,"norm")==0) {
	dPhase=0;
      }
      else {
	if(strcasecmp(sPhase,"rev")==0) {
	  dPhase=1;
	}
	else {
	  printf("agen: invalid phase setting\n");
	  exit(1);
	}
      }  
    }
    if(dMatch==0) {
      printf("agen: invalid option\n");
      exit(1);
    }
  }

  /* Convert db to ratio */
  fRatio=pow(10,(fLevel/20));

  /* Open the audio device */
  hAudio=open(sDevice,O_WRONLY);
  if(hAudio<0) {
    printf("agen: can't open audio device\n");
    exit(1);
  }

  /* Set channelization */
  dRequest=1;     /* Stereo Mode */
  ioctl(hAudio,SNDCTL_DSP_STEREO,&dRequest);
  if(dRequest!=1) {    
    printf("agen: can't configure channelization on %s\n",sDevice);
    close(hAudio);
    exit(1);
  }

  /* Set sample size */
  switch(SAMPLE_SIZE) {
  case 8:
    dRequest=AFMT_U8;
    ioctl(hAudio,SNDCTL_DSP_SETFMT,&dRequest);
    if(dRequest!=AFMT_U8) {
      printf("agen: eight bit samples not supported on %s\n",sDevice);
      close(hAudio);
      exit(1);
    }
    break;
  case 16:
    dRequest=AFMT_S16_LE;
    ioctl(hAudio,SNDCTL_DSP_SETFMT,&dRequest);
    if(dRequest!=AFMT_S16_LE) {
      printf("agen: sixteen bit samples not supported on %s\n",sDevice);
      close(hAudio);
      exit(1);
    }
    break;
  }

  /* Determine Effective Sampling Rate */
  dSampleRate=SAMPLE_RATE;
  ioctl(hAudio,SNDCTL_DSP_SPEED,&dSampleRate);
  fSampleRate=(float)dSampleRate;

  /* Display Settings (if requested) */
  if(dDebug==1) {
    printf("--Audio Generator Settings--\n");
    printf("Frequency: %5.0f Hz\n",fFreq);
    printf("Level: %3.1f dB\n",fLevel);
    printf("Channel(s): ");
    switch(dChan) {
    case 0:
      printf("BOTH\n");
      break;
    case 1:
      printf("LEFT\n");
      break;
    case 2:
      printf("RIGHT\n");
      break;
    }
    printf("Phasing: ");
    switch(dPhase) {
    case 0:
      printf("NORMAL\n");
      break;
    case 1:
      printf("REVERSE\n");
      break;
    }
    printf("Effective Sample Rate: %d samples/sec\n",dSampleRate);
  }

  /* Setup time data */
  time(&tmTimestamp);
  tmTime=tmTimestamp;
  if(dTime>0) {
    tmTimestamp+=dTime;
  }
  else {
    tmTimestamp--;
  }

  /* Output audio */

  switch(SAMPLE_SIZE) {
  case 8:
    fGain=127*fRatio;
    ldTimeline=0;
    ldLimit=BUFFER_SIZE/2;
    while(tmTimestamp!=tmTime) { 
      i=0;
      for(ldCount=ldTimeline;ldCount<ldLimit;ldCount++) {
	fAngle=2*PI*ldCount*fFreq/fSampleRate;
	if(dChan==0 || dChan==1) {
	  cBuffer[i++]=fGain*sin(fAngle)+128;
	}
	else {
	  cBuffer[i++]=128;
	}
	if(dChan==0 || dChan==2) {
	  if(dPhase==0) {
	    cBuffer[i++]=fGain*sin(fAngle)+128;
	  }
	  else {
	    cBuffer[i++]=-fGain*sin(fAngle)+128;
	  }
	}
	else {
	  cBuffer[i++]=128;
	}
      }
      write(hAudio,cBuffer,BUFFER_SIZE);
      ldLimit+=(BUFFER_SIZE/2);
      ldTimeline+=(BUFFER_SIZE/2);
      time(&tmTime);
    }  
    break;
  case 16:
    fGain=32767*fRatio;
    ldTimeline=0;
    ldLimit=BUFFER_SIZE/4;
    while(tmTimestamp!=tmTime) { 
      i=0;
      for(ldCount=ldTimeline;ldCount<ldLimit;ldCount++) {
	fAngle=2*PI*ldCount*fFreq/fSampleRate;
	dSample=fGain*sin(fAngle);
	if(dChan==0 || dChan==1) {
	  cBuffer[i++]=0xFF&dSample;
	  cBuffer[i++]=(0xFF&(dSample>>8));  
	}
	else {
	  cBuffer[i++]=0;
	  cBuffer[i++]=0;
	}
	if(dChan==0 || dChan==2) {
	  if(dPhase==1) {
	    dSample=-dSample;
	  }
	  cBuffer[i++]=0xFF&dSample;
	  cBuffer[i++]=(0xFF&(dSample>>8));
	} 
	else {
	  cBuffer[i++]=0;
	  cBuffer[i++]=0;
	}
      }
      write(hAudio,cBuffer,BUFFER_SIZE);
      ldLimit+=(BUFFER_SIZE/4);
      ldTimeline+=(BUFFER_SIZE/4);
      time(&tmTime);
    }  
  break;
  }

  /* close files and finish */
  close(hAudio);

  exit(0);
}
